File: pascals-triangle-ii.go
// 1
// 1 1
// 1 2 1
// 1 3 3 1
// 1 4 6 4 1

// ai =  
// make an array and repeatedly perform ai = ai + ai-1 --- i in [1,rowIndex] for r in [0,rowIndex]
// init conds for each row a0 = a_rowindex = 1

// rowIndex 2
// r 2
// ans [1 3 1 1]
//        i
// ans  1 3 3 1
//      i
// last 1 2 1 0
// r 2

func getRow(rowIndex int) []int {
    ans := make([]int, 1 + rowIndex)
    last := make([]int, 1 + rowIndex)
    ans[0] = 1
    last[0] = 1
    for r := range rowIndex + 1 {
        ans[r] = 1
        for i := 1; i < r; i ++ {
            ans[i] = last[i] + last[i - 1]
        }
        for i := range (len(ans)) {
            last[i] = ans[i]
        }
    }
    return ans
}

File: lc/py/pascals-triangle-ii.py
# 0: 1 0 0 0 0
# 1: 1 1 0 0 0
# 2: 1 2 1 0 0 -- 1 up-dependent change required
# 3: 1 3 3 1 0 -- 2 up-dependent changes required
# 4: 1 4 6 4 1 -- 3 changes required
# allocate a[rI+1]
# loop i for each row: 0..rI
#     ans[i] = 1
#     loop: j = rI - 1; j > 0 ; j--
#         ans[j] = ans[j] + ans[j-1]
# return ans
# ri = 4
# ans = [1 2 1 1 1]
# rI: 2
# ans [1 2 1]
#          i
from typing import List

class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        ans = [0 for _ in range(rowIndex + 1)]
        for i in range(1 + rowIndex):
            ans[i] = 1
            for j in range(i-1,0,-1):
                ans[j] += ans[j-1]
        return ans
